[SQL][SPARK-17490] Optimize SerializeFromObject() for a primitive array

kiszk opened the pull request:
## What changes were proposed in this pull request?

Waiting for merging #13680

This PR optimizes `SerializeFromObject()` for an primitive array. This is derived from #13758 to address one of problems by using a simple way in #13758.

The current implementation always generates `GenericArrayData` from `SerializeFromObject()` for any type of an array in a logical plan. This involves a boxing at a constructor of `GenericArrayData` when `SerializedFromObject()` has an primitive array.

This PR enables to generate `UnsafeArrayData` from `SerializeFromObject()` for a primitive array. It can avoid boxing to create an instance of `ArrayData` in the generated code by Catalyst.

This PR also generate `UnsafeArrayData` in a case for `RowEncoder.serializeFor` or `CatalystTypeConverters.createToCatalystConverter`.

Performance improvement of `SerializeFromObject()` is up to 2.0x

```
OpenJDK 64-Bit Server VM 1.8.0_91-b14 on Linux 4.4.11-200.fc22.x86_64
Intel Xeon E3-12xx v2 (Ivy Bridge)

Without this PR
Write an array in Dataset:               Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            556 /  608         15.1          66.3       1.0X
Double                                        1668 / 1746          5.0         198.8       0.3X

with this PR
Write an array in Dataset:               Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            352 /  401         23.8          42.0       1.0X
Double                                         821 /  885         10.2          97.9       0.4X
```

Here is an example program that will happen in mllib as described in [SPARK-16070](https://issues.apache.org/jira/browse/SPARK-16070).

```
sparkContext.parallelize(Seq(Array(1, 2)), 1).toDS.map(e => e).show
```

Generated code before applying this PR

``` java
/* 039 */   protected void processNext() throws java.io.IOException {
/* 040 */     while (inputadapter_input.hasNext()) {
/* 041 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 042 */       int[] inputadapter_value = (int[])inputadapter_row.get(0, null);
/* 043 */
/* 044 */       Object mapelements_obj = ((Expression) references[0]).eval(null);
/* 045 */       scala.Function1 mapelements_value1 = (scala.Function1) mapelements_obj;
/* 046 */
/* 047 */       boolean mapelements_isNull = false || false;
/* 048 */       int[] mapelements_value = null;
/* 049 */       if (!mapelements_isNull) {
/* 050 */         Object mapelements_funcResult = null;
/* 051 */         mapelements_funcResult = mapelements_value1.apply(inputadapter_value);
/* 052 */         if (mapelements_funcResult == null) {
/* 053 */           mapelements_isNull = true;
/* 054 */         } else {
/* 055 */           mapelements_value = (int[]) mapelements_funcResult;
/* 056 */         }
/* 057 */
/* 058 */       }
/* 059 */       mapelements_isNull = mapelements_value == null;
/* 060 */
/* 061 */       serializefromobject_argIsNulls[0] = mapelements_isNull;
/* 062 */       serializefromobject_argValue = mapelements_value;
/* 063 */
/* 064 */       boolean serializefromobject_isNull = false;
/* 065 */       for (int idx = 0; idx < 1; idx++) {
/* 066 */         if (serializefromobject_argIsNulls[idx]) { serializefromobject_isNull = true; break; }
/* 067 */       }
/* 068 */
/* 069 */       final ArrayData serializefromobject_value = serializefromobject_isNull ? null : new org.apache.spark.sql.catalyst.util.GenericArrayData(serializefromobject_argValue);
/* 070 */       serializefromobject_holder.reset();
/* 071 */
/* 072 */       serializefromobject_rowWriter.zeroOutNullBytes();
/* 073 */
/* 074 */       if (serializefromobject_isNull) {
/* 075 */         serializefromobject_rowWriter.setNullAt(0);
/* 076 */       } else {
/* 077 */         // Remember the current cursor so that we can calculate how many bytes are
/* 078 */         // written later.
/* 079 */         final int serializefromobject_tmpCursor = serializefromobject_holder.cursor;
/* 080 */
/* 081 */         if (serializefromobject_value instanceof UnsafeArrayData) {
/* 082 */           final int serializefromobject_sizeInBytes = ((UnsafeArrayData) serializefromobject_value).getSizeInBytes();
/* 083 */           // grow the global buffer before writing data.
/* 084 */           serializefromobject_holder.grow(serializefromobject_sizeInBytes);
/* 085 */           ((UnsafeArrayData) serializefromobject_value).writeToMemory(serializefromobject_holder.buffer, serializefromobject_holder.cursor);
/* 086 */           serializefromobject_holder.cursor += serializefromobject_sizeInBytes;
/* 087 */
/* 088 */         } else {
/* 089 */           final int serializefromobject_numElements = serializefromobject_value.numElements();
/* 090 */           serializefromobject_arrayWriter.initialize(serializefromobject_holder, serializefromobject_numElements, 4);
/* 091 */
/* 092 */           for (int serializefromobject_index = 0; serializefromobject_index < serializefromobject_numElements; serializefromobject_index++) {
/* 093 */             if (serializefromobject_value.isNullAt(serializefromobject_index)) {
/* 094 */               serializefromobject_arrayWriter.setNullInt(serializefromobject_index);
/* 095 */             } else {
/* 096 */               final int serializefromobject_element = serializefromobject_value.getInt(serializefromobject_index);
/* 097 */               serializefromobject_arrayWriter.write(serializefromobject_index, serializefromobject_element);
/* 098 */             }
/* 099 */           }
/* 100 */         }
/* 101 */
/* 102 */         serializefromobject_rowWriter.setOffsetAndSize(0, serializefromobject_tmpCursor, serializefromobject_holder.cursor - serializefromobject_tmpCursor);
/* 103 */       }
/* 104 */       serializefromobject_result.setTotalSize(serializefromobject_holder.totalSize());
/* 105 */       append(serializefromobject_result);
/* 106 */       if (shouldStop()) return;
/* 107 */     }
/* 108 */   }
/* 109 */ }
```

Generated code after applying this PR

``` java
/* 035 */   protected void processNext() throws java.io.IOException {
/* 036 */     while (inputadapter_input.hasNext()) {
/* 037 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 038 */       int[] inputadapter_value = (int[])inputadapter_row.get(0, null);
/* 039 */
/* 040 */       Object mapelements_obj = ((Expression) references[0]).eval(null);
/* 041 */       scala.Function1 mapelements_value1 = (scala.Function1) mapelements_obj;
/* 042 */
/* 043 */       boolean mapelements_isNull = false || false;
/* 044 */       int[] mapelements_value = null;
/* 045 */       if (!mapelements_isNull) {
/* 046 */         Object mapelements_funcResult = null;
/* 047 */         mapelements_funcResult = mapelements_value1.apply(inputadapter_value);
/* 048 */         if (mapelements_funcResult == null) {
/* 049 */           mapelements_isNull = true;
/* 050 */         } else {
/* 051 */           mapelements_value = (int[]) mapelements_funcResult;
/* 052 */         }
/* 053 */
/* 054 */       }
/* 055 */       mapelements_isNull = mapelements_value == null;
/* 056 */
/* 057 */       boolean serializefromobject_isNull = mapelements_isNull;
/* 058 */       final ArrayData serializefromobject_value = serializefromobject_isNull ? null : org.apache.spark.sql.catalyst.expressions.UnsafeArrayData.fromPrimitiveArray(mapelements_value);
/* 059 */       serializefromobject_isNull = serializefromobject_value == null;
/* 060 */       serializefromobject_holder.reset();
/* 061 */
/* 062 */       serializefromobject_rowWriter.zeroOutNullBytes();
/* 063 */
/* 064 */       if (serializefromobject_isNull) {
/* 065 */         serializefromobject_rowWriter.setNullAt(0);
/* 066 */       } else {
/* 067 */         // Remember the current cursor so that we can calculate how many bytes are
/* 068 */         // written later.
/* 069 */         final int serializefromobject_tmpCursor = serializefromobject_holder.cursor;
/* 070 */
/* 071 */         if (serializefromobject_value instanceof UnsafeArrayData) {
/* 072 */           final int serializefromobject_sizeInBytes = ((UnsafeArrayData) serializefromobject_value).getSizeInBytes();
/* 073 */           // grow the global buffer before writing data.
/* 074 */           serializefromobject_holder.grow(serializefromobject_sizeInBytes);
/* 075 */           ((UnsafeArrayData) serializefromobject_value).writeToMemory(serializefromobject_holder.buffer, serializefromobject_holder.cursor);
/* 076 */           serializefromobject_holder.cursor += serializefromobject_sizeInBytes;
/* 077 */
/* 078 */         } else {
/* 079 */           final int serializefromobject_numElements = serializefromobject_value.numElements();
/* 080 */           serializefromobject_arrayWriter.initialize(serializefromobject_holder, serializefromobject_numElements, 4);
/* 081 */
/* 082 */           for (int serializefromobject_index = 0; serializefromobject_index < serializefromobject_numElements; serializefromobject_index++) {
/* 083 */             if (serializefromobject_value.isNullAt(serializefromobject_index)) {
/* 084 */               serializefromobject_arrayWriter.setNullInt(serializefromobject_index);
/* 085 */             } else {
/* 086 */               final int serializefromobject_element = serializefromobject_value.getInt(serializefromobject_index);
/* 087 */               serializefromobject_arrayWriter.write(serializefromobject_index, serializefromobject_element);
/* 088 */             }
/* 089 */           }
/* 090 */         }
/* 091 */
/* 092 */         serializefromobject_rowWriter.setOffsetAndSize(0, serializefromobject_tmpCursor, serializefromobject_holder.cursor - serializefromobject_tmpCursor);
/* 093 */       }
/* 094 */       serializefromobject_result.setTotalSize(serializefromobject_holder.totalSize());
/* 095 */       append(serializefromobject_result);
/* 096 */       if (shouldStop()) return;
/* 097 */     }
/* 098 */   }
/* 099 */ }
```
## How was this patch tested?

Added a test in `DatasetSuite`, `RowEncoderSuite`, and `CatalystTypeConvertersSuite`

kiszk committed:
initial commit

SparkQA commented:
**[Test build #65201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65201/consoleFull)** for PR 15044 at commit [`e7c1a24`](https://github.com/apache/spark/commit/e7c1a24cc53c4554d1e0e0889b0ce394dc0134b7).

SparkQA commented:
**[Test build #65201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65201/consoleFull)** for PR 15044 at commit [`e7c1a24`](https://github.com/apache/spark/commit/e7c1a24cc53c4554d1e0e0889b0ce394dc0134b7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65201/
Test FAILed.

kiszk committed:
fix test failures

SparkQA commented:
**[Test build #65216 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65216/consoleFull)** for PR 15044 at commit [`2b22d12`](https://github.com/apache/spark/commit/2b22d128ef4c51643cd4dcdbe17a1f3d28362a90).

SparkQA commented:
**[Test build #65216 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65216/consoleFull)** for PR 15044 at commit [`2b22d12`](https://github.com/apache/spark/commit/2b22d128ef4c51643cd4dcdbe17a1f3d28362a90).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65216/
Test PASSed.

hvanhovell commented:
@kiszk this seems nice, is this ready for review?

hvanhovell commented:
retest this please

SparkQA commented:
**[Test build #66357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66357/consoleFull)** for PR 15044 at commit [`2b22d12`](https://github.com/apache/spark/commit/2b22d128ef4c51643cd4dcdbe17a1f3d28362a90).

kiszk commented:
Thanks, I should update one file and update it later today in Japan.

PR #13758 can also solve this issue without allocating UnsafeArrayData. I think that PR #13758 is a generic solution and has small amount of changes. Which PR is preferable, #15044 or #13758?

SparkQA commented:
**[Test build #66357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66357/consoleFull)** for PR 15044 at commit [`2b22d12`](https://github.com/apache/spark/commit/2b22d128ef4c51643cd4dcdbe17a1f3d28362a90).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66357/
Test FAILed.

kiszk committed:
rebase

add calling primitiveFromArray for other types

SparkQA commented:
**[Test build #66391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66391/consoleFull)** for PR 15044 at commit [`df300c5`](https://github.com/apache/spark/commit/df300c534670e65cb44ae570859711382f85043b).

SparkQA commented:
**[Test build #66391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66391/consoleFull)** for PR 15044 at commit [`df300c5`](https://github.com/apache/spark/commit/df300c534670e65cb44ae570859711382f85043b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66391/
Test PASSed.

kiszk commented:
@hvanhovell this PR is ready for review. Which PR is preferable, #15044 or #13758?

hvanhovell commented:
@kiszk this PR is. It is much more concise!

hvanhovell commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala:
```diff
@@ -0,0 +1,72 @@
+ *
+ * Benchmarks in this file are skipped in normal builds.
+ */
+class PrimitiveArrayBenchmark extends BenchmarkBase {
```
Could you post some of the results?

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -441,6 +441,28 @@ object ScalaReflection extends ScalaReflection {
           val newPath = s"""- array element class: "$clsName"""" +: walkedTypePath
           MapObjects(serializerFor(_, elementType, newPath), input, dt)
 
+         case dt @ (BooleanType | ByteType | ShortType | IntegerType | LongType |
```
I am afraid that I cannot understand your question correctly.
This case covers only cases that we are interested in (e.g. generate `UnsafeArrayData`). Other cases are covered by [here](https://github.com/apache/spark/pull/15044/files/df300c534670e65cb44ae570859711382f85043b#diff-89643554d9757dd3e91abff1cc6096c7R466).

Is this an answer to you?

hvanhovell replied:
I mean that there a few more places in which we can apply this:

- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L125-L128
- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala#L161-L172

kiszk replied:
Thank you for your clarification. Let me check them.

kiszk committed:
update benchmark program

SparkQA commented:
**[Test build #66453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66453/consoleFull)** for PR 15044 at commit [`68d34de`](https://github.com/apache/spark/commit/68d34de0eb41a1e2c225752cd4fd4e6d4cc48e7a).

SparkQA commented:
**[Test build #66453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66453/consoleFull)** for PR 15044 at commit [`68d34de`](https://github.com/apache/spark/commit/68d34de0eb41a1e2c225752cd4fd4e6d4cc48e7a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PrimitiveArrayBenchmark extends BenchmarkBase `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66453/
Test FAILed.

kiszk committed:
create UnsafeArrayData from a primitive array in CatalystTypeConverters

SparkQA commented:
**[Test build #66582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66582/consoleFull)** for PR 15044 at commit [`c10277c`](https://github.com/apache/spark/commit/c10277c825af42213e0d1afa7881a5e55dbbfd54).

kiszk committed:
create UnsafeArrayData from a primitive array in RowEncoder.serializeFor

SparkQA commented:
**[Test build #66583 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66583/consoleFull)** for PR 15044 at commit [`59fdf76`](https://github.com/apache/spark/commit/59fdf763c5eeefbf695b035edb2191802833fe6e).

SparkQA commented:
**[Test build #66582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66582/consoleFull)** for PR 15044 at commit [`c10277c`](https://github.com/apache/spark/commit/c10277c825af42213e0d1afa7881a5e55dbbfd54).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66582/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66583/
Test PASSed.

kiszk commented:
@hvanhovell, after my investigations, I have added code to generate `UnsafeArrayData' at two code paths. Could you please review this again?

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala:
```diff
@@ -61,4 +63,39 @@ class CatalystTypeConvertersSuite extends SparkFunSuite {
     assert(CatalystTypeConverters.createToCatalystConverter(IntegerType)(Some(123)) === 123)
   }
+
+  test("primitive array handing") {
```
Nit: handing => handling

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala:
```diff
@@ -61,4 +63,39 @@ class CatalystTypeConvertersSuite extends SparkFunSuite {
+      == doubleUnsafeArray)
+  }
+
+  test("An array with null handing") {
```
Nit: handing => handling

kiszk replied:
Good catch. Addressed both of them.

kiszk committed:
fix typo

SparkQA commented:
**[Test build #66941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66941/consoleFull)** for PR 15044 at commit [`eeb2747`](https://github.com/apache/spark/commit/eeb2747ef5200505f70d67880b933978d97d59d7).

SparkQA commented:
**[Test build #66941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66941/consoleFull)** for PR 15044 at commit [`eeb2747`](https://github.com/apache/spark/commit/eeb2747ef5200505f70d67880b933978d97d59d7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66941/
Test FAILed.

kiszk commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #66945 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66945/consoleFull)** for PR 15044 at commit [`eeb2747`](https://github.com/apache/spark/commit/eeb2747ef5200505f70d67880b933978d97d59d7).

SparkQA commented:
**[Test build #66945 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66945/consoleFull)** for PR 15044 at commit [`eeb2747`](https://github.com/apache/spark/commit/eeb2747ef5200505f70d67880b933978d97d59d7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66945/
Test PASSed.

kiszk commented:
@hvanhovell could you please review this again?

SparkQA commented:
**[Test build #67978 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67978/consoleFull)** for PR 15044 at commit [`8b79146`](https://github.com/apache/spark/commit/8b79146c168b0e0226af256b5d967abc6b0dcafb).

SparkQA commented:
**[Test build #67978 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67978/consoleFull)** for PR 15044 at commit [`8b79146`](https://github.com/apache/spark/commit/8b79146c168b0e0226af256b5d967abc6b0dcafb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67978/
Test FAILed.

kiszk committed:
revert changes for ValdateExternType

SPARK-18070 requests any ArrayType to accept an element-nullable array and an element-non-nullable array

SparkQA commented:
**[Test build #68054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68054/consoleFull)** for PR 15044 at commit [`b5473e3`](https://github.com/apache/spark/commit/b5473e3b6675c6bbfdd3fe261699f855faf39c94).

SparkQA commented:
**[Test build #68054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68054/consoleFull)** for PR 15044 at commit [`b5473e3`](https://github.com/apache/spark/commit/b5473e3b6675c6bbfdd3fe261699f855faf39c94).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68054/
Test FAILed.

kiszk commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #68065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68065/consoleFull)** for PR 15044 at commit [`b5473e3`](https://github.com/apache/spark/commit/b5473e3b6675c6bbfdd3fe261699f855faf39c94).

SparkQA commented:
**[Test build #68065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68065/consoleFull)** for PR 15044 at commit [`b5473e3`](https://github.com/apache/spark/commit/b5473e3b6675c6bbfdd3fe261699f855faf39c94).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68065/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -119,18 +119,36 @@ object RowEncoder {
+      val cls = inputObject.dataType.asInstanceOf[ObjectType].cls
+      et match {
+/*
+        case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType
```
What is going on here?

kiszk replied:
Sorry, my mistake

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -193,7 +211,8 @@ object RowEncoder {
     case _: DecimalType => ObjectType(classOf[java.lang.Object])
     // In order to support both Array and Seq in external row, we make this as java.lang.Object.
-    case _: ArrayType => ObjectType(classOf[java.lang.Object])
+    case a @ ArrayType(et, cn) =>
```
Why this change?

kiszk replied:
I reverted it, but it was not the same as the previous one. thanks.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:
```diff
@@ -993,7 +993,7 @@ case class ValidateExternalType(child: Expression, expected: DataType)
         Seq(classOf[java.math.BigDecimal], classOf[scala.math.BigDecimal], classOf[Decimal])
           .map(cls => s"$obj instanceof ${cls.getName}").mkString(" || ")
-      case _: ArrayType =>
+      case a @ ArrayType(et, cn) =>
```
Same here: why this change?

kiszk replied:
ditto

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:
```diff
@@ -158,6 +158,20 @@ object CatalystTypeConverters {
+        case a: Array[Float] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
+        case a: Array[Double] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
```
do we need to do this? The `CatalystTypeConverter` will be removed and replaced by encoder eventually. Does your benchmark cover this branch?

kiszk replied:
My benchmark does not cover this branch.
[This test](https://github.com/kiszk/spark/blob/b5473e3b6675c6bbfdd3fe261699f855faf39c94/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala#L72-L73) causes an failure without this branch. Should we drop this test and this branch?

cloud-fan replied:
yea let's remove them, `CatalystTypeConverter` will not be in critical path and we don't need to optimize it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -441,6 +441,28 @@ object ScalaReflection extends ScalaReflection {
+              cls.isAssignableFrom(classOf[Array[Long]]) ||
+              cls.isAssignableFrom(classOf[Array[Float]]) ||
+              cls.isAssignableFrom(classOf[Array[Double]])) {
+            StaticInvoke(
```
we can simplify it 
```
if (cls.isArray && cls.getComponentType.isPrimitive) ...
```
```

kiszk replied:
thanks, good catch.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -119,18 +119,36 @@ object RowEncoder {
-    case t @ ArrayType(et, _) => et match {
-      case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType =>
-        // TODO: validate input type for primitive array.
-        NewInstance(
```
can we do the same thing here? i.e. special handling primitive array. I know we don't have the class information here, bu we can do it in the runtime:
```
object ArrayData {
  def toArrayData(input: Any): ArrayData = input match {
    case a: Array[Boolean] => UnsafeArrayData.fromPrimitive(a)
    ...
    case other => new GenericArrayData(other)
  }
}
```

kiszk replied:
Good idea. It works well.

kiszk committed:
addressed review comments

SparkQA commented:
**[Test build #68127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68127/consoleFull)** for PR 15044 at commit [`c5378f9`](https://github.com/apache/spark/commit/c5378f95aa8df1eda8d4694bfb14a176f2004477).

SparkQA commented:
**[Test build #68127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68127/consoleFull)** for PR 15044 at commit [`c5378f9`](https://github.com/apache/spark/commit/c5378f95aa8df1eda8d4694bfb14a176f2004477).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68127/
Test PASSed.

kiszk committed:
addressed review comments

SparkQA commented:
**[Test build #68197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68197/consoleFull)** for PR 15044 at commit [`d9e5b4f`](https://github.com/apache/spark/commit/d9e5b4f9c934d1a409d9a74cb8f77745f8e46f51).

SparkQA commented:
**[Test build #68197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68197/consoleFull)** for PR 15044 at commit [`d9e5b4f`](https://github.com/apache/spark/commit/d9e5b4f9c934d1a409d9a74cb8f77745f8e46f51).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68197/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -119,18 +119,20 @@ object RowEncoder {
+        case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType =>
+          StaticInvoke(
+            classOf[ArrayData],
+            ObjectType(classOf[ArrayData]),
```
why not just use the given ArrayType `t` here?

kiszk replied:
yes, done

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -119,18 +119,20 @@ object RowEncoder {
-        ObjectType(classOf[Object]))
-    }
+    case t @ ArrayType(et, cn) =>
+      val cls = inputObject.dataType.asInstanceOf[ObjectType].cls
```
where do we use the `cls`?

kiszk replied:
oh, removed

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala:
```diff
@@ -191,6 +191,32 @@ class RowEncoderSuite extends SparkFunSuite {
     assert(encoder.serializer.head.nullable == false)
   }
 
+  test("RowEncoder should support a primitive array") {
```
nit: `RowEncoder should support primitive arrays`

kiszk replied:
done

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -987,6 +987,24 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
     }
   }
+
+  test("array") {
```
can you think of a better test name which can describe what we are testing here?

kiszk replied:
good catch, gave a better name

kiszk committed:
addressed comments

SparkQA commented:
**[Test build #68253 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68253/consoleFull)** for PR 15044 at commit [`d507cfc`](https://github.com/apache/spark/commit/d507cfce431bdee672a9dd81c5111d768cb28185).

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala:
```diff
@@ -0,0 +1,81 @@
+    val sc = sparkSession.sparkContext
+    val primitiveIntArray = Array.fill[Int](count)(65535)
+    val dsInt = sc.parallelize(Seq(primitiveIntArray), 1).toDS
+    dsInt.count
```
Want to force to build dataset

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala:
```diff
@@ -0,0 +1,81 @@
+    val intArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        dsInt.map(e => e).collect
```
I see. done

SparkQA commented:
**[Test build #68253 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68253/consoleFull)** for PR 15044 at commit [`d507cfc`](https://github.com/apache/spark/commit/d507cfce431bdee672a9dd81c5111d768cb28185).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68253/
Test PASSed.

cloud-fan commented:
LGTM, can you also address https://github.com/apache/spark/pull/15044#discussion_r86665473?

kiszk committed:
addressed review comments

kiszk committed:
addressed review comments

SparkQA commented:
**[Test build #68275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68275/consoleFull)** for PR 15044 at commit [`c217395`](https://github.com/apache/spark/commit/c2173950f399e910eb7717e2ec8f9e895e9cae1c).

kiszk commented:
Sorry for missing to commit [this](https://github.com/kiszk/spark/blob/4c679b59b34ea01474baa5d4ac4ce89133a1720d/sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala#L991).

SparkQA commented:
**[Test build #68276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68276/consoleFull)** for PR 15044 at commit [`4c679b5`](https://github.com/apache/spark/commit/4c679b59b34ea01474baa5d4ac4ce89133a1720d).

SparkQA commented:
**[Test build #68275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68275/consoleFull)** for PR 15044 at commit [`c217395`](https://github.com/apache/spark/commit/c2173950f399e910eb7717e2ec8f9e895e9cae1c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68275/
Test PASSed.

SparkQA commented:
**[Test build #68276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68276/consoleFull)** for PR 15044 at commit [`4c679b5`](https://github.com/apache/spark/commit/4c679b59b34ea01474baa5d4ac4ce89133a1720d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68276/
Test PASSed.

hvanhovell commented:
Merging to master/2.1. Thanks for the hard work!
